File: 250504/2N개 중에 N개의 숫자를 적절하게 고르기/choose-n-out-of-2n-properly.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

int n;
int num[20];
vector<int> A;
vector<int> B;
int result = INT_MAX;

int calc(){
    int a = 0;
    int b = 0;
    for (int i = 0; i < n; i++) {
        a += A[i];
        b += B[i];
    }
    return abs(a - b);
}

void func(int idx) {
    if (A.size() == n && B.size() == n) {
        int ans = calc();
        //cout << "heY ~ " << ans << '\n';
        result = min(result, ans);
        return;
    }
    if (idx == 2 * n) return;

    if (A.size() < n) {
        A.push_back(num[idx]);
        func(idx + 1);
        A.pop_back();
    }

    if (B.size() < n) {
        B.push_back(num[idx]);
        func(idx + 1);
        B.pop_back();
    }
}

int main() {
    cin >> n;
    for (int i = 0; i < 2 * n; i++) {
        cin >> num[i];
    }
    func(0);

    cout << result;
    return 0;
}

File: 250810/2N개 중에 N개의 숫자를 적절하게 고르기/choose-n-out-of-2n-properly.py
from math import inf
from collections import Counter

n = int(input())
num = list(map(int, input().split()))
ans = inf

selected = []

def calc():
    # selected에 있는 원소들을 num에서 '개수만큼' 제거한 나머지가 remain
    cnt = Counter(selected)
    remain = []
    for x in num:
        if cnt[x] > 0:
            cnt[x] -= 1
        else:
            remain.append(x)

    # 예시 목적함수: 선택합과 남은합의 차이의 절댓값
    return abs(sum(remain) - sum(selected))

def func(idx, picked):
    global ans
    # 가지치기: 남은 원소로 n개 못 채우면 중단
    if picked + (2 * n - idx) < n:
        return

    if picked == n:
        ans = min(ans, calc())
        # 디버깅 출력 원하면:
        # if ans in (7, 19):
        #     print(*selected)
        return

    if idx == 2 * n:
        return

    # 선택
    selected.append(num[idx])
    func(idx + 1, picked + 1)
    selected.pop()

    # 미선택
    func(idx + 1, picked)
    
func(0, 0)
print(ans)